warehouse/query-core/src/main/java/datawave/query/iterator/waitwindow/WaitWindowObserver.java

```diff
@@ -0,0 +1,584 @@
+                }
+            }
+            newKey = new Key(key.getRow(), key.getColumnFamily(), colQual);
+        }
+        return ensureYieldKeyAfterRangeStart(newKey);
+    }
+
+    /*
+     * When the current seekRange is non-inclusive, we can not return the startKey of the range as a yield key. Instead, we have to return the following key. We
+     * also want to ensure that either the YIELD_AT_BEGIN or YIELD_AT_END markers are included in the yield key
+     */
+    private Key ensureYieldKeyAfterRangeStart(Key key) {
+        if (!this.seekRange.isStartKeyInclusive()) {
+            Key seekStartKey = this.seekRange.getStartKey();
+            boolean isShardKey = isShardKey(seekStartKey);
+            if (isShardKey && key.compareTo(seekStartKey, PartialKey.ROW_COLFAM) <= 0) {
+                // shard range
+                Key followingKey = seekStartKey.followingKey(PartialKey.ROW_COLFAM);
+                if (hasMarker(followingKey)) {
+                    return followingKey;
+                } else {
+                    // followingKey already appended a null character to the colFam, so it's safe to append "\uffffYIELD_AT_END"
+                    Text colFam = new Text(followingKey.getColumnFamily().toString() + YIELD_AT_END_STR);
+                    return new Key(followingKey.getRow(), colFam);
+                }
+            }
+            if (!isShardKey && key.compareTo(seekStartKey, PartialKey.ROW_COLFAM_COLQUAL) <= 0) {
+                // document range
+                Key followingKey = seekStartKey.followingKey(PartialKey.ROW_COLFAM_COLQUAL);
+                if (hasMarker(followingKey)) {
+                    return followingKey;
+                } else {
+                    // followingKey already appended a null character to the colQual, so it's safe to append "\uffffYIELD_AT_END"
+                    Text colQual = new Text(followingKey.getColumnQualifier().toString() + YIELD_AT_END_STR);
+                    return new Key(followingKey.getRow(), followingKey.getColumnFamily(), colQual);
+                }
+            }
+        }
+        return key;
+    }
+
+    /*
+     * When yieldKey is set and collectTimingDetails=true, then we yield on the second call from QueryIterator.hasTop so that the first call can return the
+     * document that contains the WAIT_WINDOW_OVERRUN and TIMING_METADATA attributes. Since we can not yield on the same key that was just returned, we have to
+     * yield on a key that follows the yieldKey. For YIELD_AT_END, we add a null char after YIELD_AT_END which is at the end of the colFam or colQual. For
+     * YIELD_AT_BEGIN, we add a null character after that marker.
+     */
+    private Key yieldKeyAfterOverrun(Key key) {
+        Text row = key.getRow();
+        Text colFam = key.getColumnFamily();
+        Text colQual = key.getColumnQualifier();
+        if (hasBeginMarker(key)) {
+            if (hasBeginMarker(colFam)) {
```

Comment:
I am trying to figure out why we can have a marker in wither the colFam or the colQual.  Why not always place in the colQual regardless?  I believe you have have a key with an empty (0 length) colFam and still have data in the colQual.   If that presumpting is incorrect then this makes perfect sense to me.  If that presumtion is correct, then my question stands.

Reply:
nevermind... explained in the createDocumentYieldKey method above.